### PR TITLE
Add pallet-sense

### PR DESCRIPTION
**Pull request by vovacha**
`sense-pallet` migration.

**What was done?**
- [x] Change the syntax to the newest Substrate version with minor changes.
- [x] Add tests and mocks.
- [x] Add benchmarks.

**Tests:**
```
❯ cargo test -p pallet-sense --all-features
   Compiling pallet-sense v0.1.0 (/Users/vovacha/Projects/gamedao/gamedao-acala/gamedao/sense)
    Finished test [unoptimized + debuginfo] target(s) in 2.71s
     Running unittests (target/debug/deps/pallet_sense-952d6876c90a0174)

running 6 tests
test mock::__construct_runtime_integrity_test::runtime_integrity_tests ... ok
test tests::sense_mod_rep ... ok
test tests::sense_mod_trust ... ok
test tests::sense_mod_xp ... ok
test tests::sense_create_entity ... ok
test benchmarking::benchmark_tests::test_benchmarks ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests pallet-sense

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

### CODE REVIEW

File: sense/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+				Sense::<Test>::insert(
+					account, Entity::new(account, block_number, 0, vec![1,2,3])
+				);
+				$storage::<Test>::insert(
+					account, EntityProperty::new(account, block_number)
+				);
+		
+				assert_ok!($extrinsic(
+					RawOrigin::Root.into(), account, 125)
+				);
```
Comment (bigkon) [inline sense/src/tests.rs:94]: What about extending this test to also check next items?

1. StorageMap value updated after calling extrinsic (`SenseXP` etc.)
2. Event is generated (`EntityMutateXP` etc.)

File: sense/src/tests.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+		assert_eq!(
+			System::events(),
+			vec![EventRecord {
+				phase: Phase::Initialization,
+				event: Event::ZeroSense(ZeroEvent::EntityInit(account, block_number)),
+				topics: vec![],
+			}]
+		);
+
```
Comment (bigkon) [inline sense/src/tests.rs:57]: Check `Nonce` value increased in storage as a result of successful extrinsic call.

**Review by bigkon — CHANGES_REQUESTED** [review]

File: sense/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+		#[pallet::weight(<T as Config>::WeightInfo::mod_rep())]
+		pub fn mod_rep(origin: OriginFor<T>, account: T::AccountId, value: u8) -> DispatchResult {
+
+			ensure_root(origin)?;
+			ensure!( <Sense<T>>::contains_key(&account), Error::<T>::EntityUnknown );
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			let v = u64::from(value);
+			let current = Self::xp(&account);
```
[the remark this answers is not in the document]
Comment (vovacha) [reply sense/src/lib.rs:214]: Oh, nice!

File: sense/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+		#[pallet::weight(<T as Config>::WeightInfo::mod_trust())]
+		pub fn mod_trust(origin: OriginFor<T>, account: T::AccountId, value: u8) -> DispatchResult {
+
+			ensure_root(origin)?;
+			ensure!( <Sense<T>>::contains_key(&account), Error::<T>::EntityUnknown );
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			let v = u64::from(value);
+			let current = Self::xp(&account);
```
[the remark this answers is not in the document]
Comment (vovacha) [reply sense/src/lib.rs]: Fixed.

File: sense/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vovacha) [reply sense/src/lib.rs]: Fixed.

File: sense/src/tests.rs (hunk 2 above)
Comment (vovacha) [reply sense/src/tests.rs:57]: I've added it as a TODO item for now.

File: sense/src/tests.rs (hunk 1 above)
Comment (vovacha) [reply sense/src/tests.rs:97]: I've added it as a TODO item for now.

File: sense/src/mock.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+impl pallet_sense::Config for Test {
+	type Event = Event;
+	type ForceOrigin = frame_system::EnsureRoot<Self::AccountId>;
+	type WeightInfo = ();
+}
+
+// Build genesis storage according to the mock runtime.
+pub fn new_test_ext() -> sp_io::TestExternalities {
+	system::GenesisConfig::default().build_storage::<Test>().unwrap().into()
```
[the remark this answers is not in the document]
Comment (vovacha) [reply sense/src/mock.rs:67]: I think it'll be an implicit behavior. It's also used for benchmarks and some of the future tests could possibly have another value.

**Review by bigkon — APPROVED** [review]

**Comment by 2075** [comment]
nice review, thanks @vovacha @bigkon

**Review by 2075 — APPROVED** [review]

**Comment by 2075** [comment]
we can also merge with two reviewers, if you like, @cryply can review still.

**Pull request merged**